src/widget_chooser.c

```diff
@@ -0,0 +1,502 @@
+	}
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Removeselected                                                      *
+ ***********************************************************************/
+
+void widget_chooser_removeselected(variable *var)
+{
+	gchar            *var1;
+	gint              var2;
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	fprintf(stderr, "%s(): Removeselected not implemented for this widget.\n",
+		__func__);
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Save                                                                *
+ ***********************************************************************/
+
+void widget_chooser_save(variable *var)
+{
+	FILE             *outfile;
+	GList            *element;
+	GtkTextBuffer    *buffer;
+	GtkTextIter       start, end;
+	gchar            *act;
+	gchar            *filename = NULL;
+	gchar            *text;
+	gchar            *string;
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	/* We'll use the output file filename if available */
+	act = attributeset_get_first(&element, var->Attributes, ATTR_OUTPUT);
+	while (act) {
+		if (strncasecmp(act, "file:", 5) == 0 && strlen(act) > 5) {
+			filename = act + 5;
+			break;
+		}
+		act = attributeset_get_next(&element, var->Attributes, ATTR_OUTPUT);
+	}
+
+	/* If we have a valid filename then open it and dump the widget's data to it */
+	if (filename) {
+		text = gtk_file_chooser_get_filename(GTK_FILE_CHOOSER(var->Widget));
+			if (text) {
+				string = g_strdup(text);
+				if ((outfile = fopen(filename, "w"))) {
+					fprintf(outfile, "%s", string);
+					fclose(outfile);
+				} else {
+					fprintf(stderr, "%s(): Couldn't open '%s' for writing.\n",
+						__func__, filename);
+				}
+				g_free(text);
+			}
+	} else {
+		fprintf(stderr, "%s(): No <output file> directive found.\n", __func__);
+	}
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Input by Command                                                    *
+ ***********************************************************************/
+
+static void widget_chooser_input_by_command(variable *var, char *command)
+{
+	FILE             *infile;
+	GString          *text = g_string_sized_new(512);
+	gchar             line[512];
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+#ifdef DEBUG_CONTENT
+	fprintf(stderr, "%s(): command: '%s'\n", __func__, command);
+#endif
+
+	/* Opening pipe for reading... */
+	if (infile = widget_opencommand(command)) {
+		/* Read the file one line at a time */
+		while (fgets(line, 512, infile)) {
```

Comment:
This 512 is dangerous - better use `sizeof(line)` in case someone increases the buffer size but forgets to change here.

Reply:
Buffer resizing is taken care of by `g_string_append` in the next line, AFAIK.

Reply:
I'm talking about `line`, not the dynamically-allocated string you're appending to. The number 512 appears in two places, and if one changes the first, one must also change the other. If you pass `sizeof(line)` to `fgets()`, the buffer size is defined only in one place.

Reply:
I misunderstood you.  You're right, I could have used `sizeof`. I'll add it to my TODO for a future PR.